### ISSUE TITLE
PhET-iO Recording does not record ruler moved by keyboard

### ISSUE THREAD

**Issue by KatieWoe** [phetsims/gravity-force-lab#242]
For https://github.com/phetsims/QA/issues/471
Win 10 Chrome on Recording wrapper (not metacog). If you use keyboard nav to pick up and move the ruler, it will look as if the ruler was not picked up and moved, just highlighted for an amount of time. Moving the robots or changing radio buttons does seem to work. Moving the ruler by hand seems to work.

**Comment by KatieWoe** [comment]
Same thing happens in mirror inputs and is easier to see there. (Doesn't happen in state that I can see.) 
![morewierdness](https://user-images.githubusercontent.com/41024075/73487368-12d59800-4364-11ea-93a2-fa98aca1c648.gif)

**Comment by KatieWoe** [comment]
Do not see this on Win 10 Firefox.

**Comment by KatieWoe** [comment]
Happens on Edge (probably because Edge is now using Chromium)

**Comment by zepumph** [comment]
This is a bug with how `GrabDragInteraction` is manipulating "input events".

Side note: @KatieWoe, adding `?phetioDebug` to wrappers runs the debug version sim with assertions in it. That could be a nice tool for QA to have. For example this error caused an assertion error, but only with `?phetioDebug`. Is that something to add to the QA book?

**Comment by zepumph** [comment]
I worked on this for a while and I'm pretty stuck. There are three-ish things going on here:

1. The immediate problem is that there is an assertion in FocusOverlay that you don't get a focus highlight if you are not focused. I updated that to only effect display that are interactive (since the downstream sim is set to active: false in the sim. This feels good to me, but @jessegreenberg please review the above commit. Do we also want to do this for Accessibility.focus() or something like that?
2. (unsure exactly about this one) `GrabDragInteraction` programmatically triggers "playback" input events from its code, so I think that in the downstream sim we are playing some events twice (once from the sim, and once from the next playback event given to it. I was able to get the grabDragInteraction working in a hackish way that I will never commit. In the below patch I added support to have nested playback events not be played back. This assumes bad bad things, but in the GrabDragInteraction it means that the nested playback events aren't tried to play back again after the the sim already did that logic (downstream from the previous playback event). I'm not how to solve this issue. I think I need reinforcements, likely from the PhET-iO side.

<details>

```diff
Index: js/dataStream.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/dataStream.js	(revision afa18dc81d5054235cafe4063b470fc38ac1be52)
+++ js/dataStream.js	(date 1580438753551)
@@ -41,6 +41,9 @@
       // even if high frequency events are filtered out.
       this.isEmittingHighFrequencyEventStack = [];
 
+      // @private {boolean[]} -
+      this.isEmittingPlaybackEventStack = [];
+
       // If the user specified log=console, then output to the console.  This is used for debugging purposes, to make
       // sure the events are accurate.
       if ( phet.phetio.queryParameters.phetioConsoleLog === 'json' ) {
@@ -214,10 +217,15 @@
 
       this.index++;
       last.children.push( event );
+      if ( event.metadata && event.metadata.playback && this.isEmittingFromPlaybackEvent() ) {
+        event.metadata.playback = false;
+      }
 
       this.stack.push( event );
       this.isEmittingHighFrequencyEventStack.push( isHighFrequencyEvent );
+      this.isEmittingPlaybackEventStack.push( event.metadata && event.metadata.playback );
       assert && assert( this.stack.length === this.isEmittingHighFrequencyEventStack.length );
+      assert && assert( this.stack.length === this.isEmittingPlaybackEventStack.length );
 
       return event.index;
     }
@@ -245,6 +253,8 @@
 
       const event = this.stack.pop();
       this.isEmittingHighFrequencyEventStack.pop();
+      this.isEmittingPlaybackEventStack.pop();
+      assert && assert( this.stack.length === this.isEmittingPlaybackEventStack.length );
       assert && assert( this.stack.length === this.isEmittingHighFrequencyEventStack.length );
 
       phet.log && phet.log( this.indent() + 'POP: ' + index );
@@ -265,6 +275,10 @@
       return _.some( this.isEmittingHighFrequencyEventStack, isHighFrequency => !isHighFrequency );
     }
 
+    isEmittingFromPlaybackEvent() {
+      return _.some( this.isEmittingPlaybackEventStack, _.identity );
+    }
+
     /**
      * Events are buffered until listeners (if any) are wired up.  Launch is called immediately and synchronously
      * for ?phetioStandalone, or after listeners are wired up if not launching with ?phetioStandalone.  Hence this

```
</details>

3. I'm not sure how far to take this issue, because I don't know how much we want to support this combo. It would be nice to support alternative input in PhET-iO, even if we don't do descriptions just yet.

**Comment by jessegreenberg** [comment]
For 1) > I updated that to only effect display that are interactive (since the downstream sim is set to active: false in the sim.

Actually, does this work? This code continues  to run in the downstream sim and so this assertion seems important. I am wondering if AccessiblePeer.isFocused needs to changed so that it doesn't use document.activeElement but uses a scenery model. Otherwise all usages of Accessibility.isFocused won't work in the downstream sim. The above commit does this and has aqua passing. @zepumph what do you think.

For 3) I agree and I am uncertain. Maybe we should put this on-hold and come back to this for Greenhouse when we have more time?

**Comment by jessegreenberg** [comment]
This change broke unit tests because `Accessibility.isFocused` now requires the A11yPointer to update `Display.focusProperty`, which requires `Display.initializeEvents`. I still think this change is OK because now `Accessibility.isFocused` will match `Display.focusProperty`.

**Comment by KatieWoe** [comment]
Applies to metacog recordings too

**Comment by zepumph** [comment]
@jessegreenberg while your commit get's past the assertion, the grabDragInteraction is not yet working for keyboard mirror inputs in chrome. Let's discuss this more during tomorrow's meeting. 

I think part of the issue is how we are focusing the draggable when the grabbable is clicked. We may need to manage both sides of focus based on interactivity.

Looking forward to discussion!

**Comment by zepumph** [comment]
I worked on this with @jessegreenberg today, and though we feel like we need to factor out usages of `activeElement` to instead use `Display.focusProperty` (separate issue to come), we don't this that is the direct problem here. Our best guess is that there are nested plaback events that are being pressed twice. 

We looked into what the scenery events look like for the GrabDragInteraction (ruler) in firefox and chrome, and they are different. On chrome there are two more blur events. The GrabDragInteraction is set up to release the draggable on blur.

This is likely the root cause of the problem, since the playback wrapper is playing back the `click` event on grab. This then triggers children events to change the focus over to the draggable. Then those focus events are played again through the phet-io playback engine as top level playback events, even though they were actually children of the `click` event and don't need to be called again as events needed for playback.

From this I think we should go back to investigating (2) from https://github.com/phetsims/gravity-force-lab/issues/242#issuecomment-580562370 with the phet-io team.

**Comment by zepumph** [comment]
After dragging @chrisklus and @samreid through the depths of many bad solutions, we finally came across an algorithm to make this work.

First, the problem at hand:
The general issue is that alternative input events coming from the PDOM break a foundational assumption that PhET-iO had on scenery inputs. This is that they are always "top level" events and cannot be triggered by other sim code. For example a "mouse down" is never something that can be called from other code, it can only come from the user. PDOM scenery events are different, sometimes listeners will blur/focus other events etc.

To solve this we developed a way for the data stream to be told to not release playback events as "playbackable". In other words to override an event's "playback" value to be false.  Our first test case was to add this logic directly into `grabDragInteraction`. This worked well, and then helped us see the general case, which is that all alternative input PDOM listeners should be surrounded with logic that tells data stream not to count downstream events as playbackable. 

For 6 hours of work, the commit is surprisingly simple, and fixed the immediate problem as well as future proofing us by providing a way to have better granularity and control over what events should be played back. 

While working on this we found a few other bugs:
- [x] `PhetioObject.phetioEventMetadata`, though an option, is mutable by parents that the option is passed to, we decided to document it as this instead of creating many new objects.
- [ ] `Input.validatePointersAction` does not need to be phetioPlayback:true since its only usage occurs during the stepSimulation event.


@samreid will you please review the commit above, and let me know what you think? You were with me until the very end, but I would like your insight on the final product.

**Comment by samreid** [comment]
I made minor documentation updates, everything else looks great, nice work!  I tested gravity-and-orbits in the playback wrapper and couldn't break it.  Closing.

**Comment by samreid** [comment]
Two questions:

a) Do you think it's unnecessary to wrap the other Input.js functions with this push/pop?  I think it's unnecessary but wanted to run it past you.
b) Could we use this new facet to simplify `handlePlaybackEvent` and its dealing with nested events on simulationStarted?

**Comment by zepumph** [comment]
>a) Do you think it's unnecessary to wrap the other Input.js functions with this push/pop? I think it's unnecessary but wanted to run it past you.

Hmm. That is a good question. Remember while testing we found that there was a blur event nested under a mouse event? Wouldn't that blur get hit twice? What would hurt by adding it to all places? As we increase our support for multiple modalities, I wouldn't be surprised if we ran into weird cases in the future. That said I don't think we need to support that now, and it may be best to hold off until we understand the problems better, instead of adding complexity without test cases.

>b) Could we use this new facet to simplify handlePlaybackEvent and its dealing with nested events on simulationStarted?

We didn't end up committing any code that will assert that nesting playback doesn't happen but the case involving `disallowPlaybackCount`. Should we? I don't think so. 

I'm not sure exactly how this code can help us with the simulation started event. We still want the children events to be playbackable. 

We ended on a system that allows playback events at any hierarchy, and it is up to the developer to make sure that events aren't being played twice. Is there a way to offload some of this labor to assertions?

No matter. I don't think any of this is blocking publication for GFL, and I will go ahead and cherry-pick the above commits to that branch and have things tested there. If you think more work is needed. Can you make other issues that point here? Or tell me what issues you think should be made and I'll be happy to.

**Comment by zepumph** [comment]
I did not cherry pick the following commits:
https://github.com/phetsims/scenery-phet/commit/d8270ac9df1cf0ac17c30e11b27df728805b6e4f
https://github.com/phetsims/scenery/commit/29656a7482cee60637113a87dd88d2c19436773c
https://github.com/phetsims/scenery/commit/603c857be1acb74ab825850e97eef91050ebeb10

And things are working well.

> If you think more work is needed. Can you make other issues that point here? Or tell me what issues you think should be made and I'll be happy to.

@samreid please open new issues for any other work you think needs to be done here. 

QA please don't close this one, but keep it assigned to @samreid when confirmed for the RC branch.

**Comment by samreid** [comment]
Everything looks good here, thanks!  No need for side issues that I can think of.  Closing.

**Comment by zepumph** [comment]
Thanks! Leaving open for QA to close when it is confirmed fixed in the GFL branch.

**Comment by samreid** [comment]
Thanks!

**Comment by loganbraywork** [comment]
Appears fixed
